src/Bicep.Core/Semantics/SymbolicReferenceCollector.cs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using System.Linq;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.Semantics;
+
+internal class SymbolicReferenceCollector : AstVisitor
+{
+    private readonly HashSet<DeclaredSymbol> symbolsReferenced = new();
+    private readonly IBinder binder;
+
+    private SymbolicReferenceCollector(IBinder binder)
+    {
+        this.binder = binder;
+    }
+
+    internal static IEnumerable<DeclaredSymbol> CollectSymbolsReferenced(IBinder binder, DeclaredSymbol symbol)
+    {
+        SymbolicReferenceCollector collector = new(binder);
+        symbol.DeclaringSyntax.Accept(collector);
+        return collector.symbolsReferenced;
+    }
+
+    internal static IEnumerable<DeclaredSymbol> CollectSymbolsReferencedRecursive(IBinder binder, DeclaredSymbol symbol)
+    {
+        Queue<DeclaredSymbol> searchQueue = new(new[] { symbol });
+        HashSet<DeclaredSymbol> searched = new();
+        SymbolicReferenceCollector collector = new(binder);
+
+        while (searchQueue.TryDequeue(out var toSearch))
+        {
+            searched.Add(toSearch);
+            toSearch.DeclaringSyntax.Accept(collector);
+
+            foreach (var referenced in collector.symbolsReferenced.Where(r => !searched.Contains(r)))
+            {
+                searchQueue.Enqueue(referenced);
+            }
+        }
```

Comment:
Because this is called independently whenever we see an `@export()`, there's probably going to be a fair amount of redundant work here in exported variables referencing each other.

That said, I'm not sure it's worth optimizing currently if it's going to complicate the code.

Reply:
The redundant work could be avoided by attaching this data to the Binder, but that approach would create more work in templates that don't export any variable (or only export one or two while defining many more non-exported variables).

Reply:
I moved the closure calculation to the Binder to eliminate the redundant work when one exported variable references another. The calculation is performed lazily, so there shouldn't be any additional overhead for templates that don't export variables.